File: 11_String/sortVowels.cpp
#include<iostream>
#include <vector>
using namespace std;

class Solution{
    public:
        string sortVowels(string s){
            vector<int> lower(26, 0);
            vector<int> upper(26, 0);

            for (int i = 0; i < s.size(); i++){
                if(s[i] == 'a' || s[i] == 'e' || 
                    s[i] == 'o' || s[i] == 'u' ||
                    s[i] == 'i'){
                        lower[s[i] - 'a']++;
                        s[i] = '#';
                }
                else if(s[i] == 'A' || s[i] == 'E' || s[i] == 'O' || s[i] == 'U' ||s[i] == 'I'){
                        upper[s[i] - 'a']++;
                        s[i] = '#';
                }
            }

            string vowel;
            for (int i = 0; i < 26; i++){
                char c = 'A' + i;
                while(upper[i] > 0){
                    vowel += c;
                    upper[i]--;
                }
            }

            for (int i = 0; i < 26; i++){
                char c = 'a' - i;
                while (lower[i] > 0){
                    vowel += c;
                    lower[i]--;
                }
            }

            int idx = 0;
            for(int i = 0; i < s.size(); i++){
                if (s[i] == '#'){
                    s[i] = vowel[idx++];
                }
            }

            return s;
        }
};

int main(){
    Solution sol;

    string s1 = "lEetcOde";
    cout << sol.sortVowels(s1) << endl;

    string s2 = "lYmpH";
    cout << sol.sortVowels(s2) << endl;

    return 0;
}

/*
-------- output -----------

➜  11_String git:(main) ✗ g++ sortVowels.cpp -o sortVowels
➜  11_String git:(main) ✗ ./sortVowels 
la]tc]dW
lYmpH
➜  11_String git:(main) ✗ 

*/

File: 11_String/string_java/sortVowels.java
import java.util.*;

public class sortVowels {
    public String convertSortVowels(String s) {
        int[] lower = new int[26];
        int[] upper = new int[26];

        char[] arr = s.toCharArray();

        // Step 1: Count vowels
        for (int i = 0; i < arr.length; i++) {
            char c = arr[i];

            if (c == 'a' || c == 'e' || c == 'i' || 
                c == 'o' || c == 'u') {
                lower[c - 'a']++;
                arr[i] = '#';
            } 
            else if (c == 'A' || c == 'E' || c == 'I' || 
                     c == 'O' || c == 'U') {
                upper[c - 'A']++;
                arr[i] = '#';
            }
        }

        // Step 2: Build sorted vowels
        StringBuilder vowels = new StringBuilder();

        // Uppercase first
        for (int i = 0; i < 26; i++) {
            while (upper[i]-- > 0) {
                vowels.append((char) ('A' + i));
            }
        }

        // Lowercase next
        for (int i = 0; i < 26; i++) {
            while (lower[i]-- > 0) {
                vowels.append((char) ('a' + i));
            }
        }

        // Step 3: Replace vowels
        int idx = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == '#') {
                arr[i] = vowels.charAt(idx++);
            }
        }

        return new String(arr);
    }
    
    public static void main(String[] args) {
        sortVowels sol = new sortVowels();

        String s1 = "lEetcde";
        System.out.println(sol.convertSortVowels(s1));

        String s2 = "lYmpH";
        System.out.println(sol.convertSortVowels(s2));
    }    
}


// import java.util.*;

// public class sortVowels {
//     public String convertSortVowels(String s){
//         Set<Character> vowelsSet = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));

//         List<Character> vowels = new ArrayList<>();

//         for(char c : s.toCharArray()){
//             if (vowelsSet.contains(c)){
//                 vowels.add(c);
//             }
//         }

//         Collections.sort(vowels);
//         StringBuffer result = new StringBuffer();
//         int index = 0;

//         for (char c : s.toCharArray()){
//             if (vowelsSet.contains(c)){
//                 result.append(vowels.get(index++));
//             } else{
//                 result.append(c);
//             }
//         }

//         return result.toString();
//     }    

//     public static void main(String[] args) {
//         sortVowels sol = new sortVowels();

//         String s1 = "lEetcde";
//         System.out.println(sol.convertSortVowels(s1));

//         String s2 = "lYmpH";
//         System.out.println(sol.convertSortVowels(s2));
//     }    
// }
